File: frontend/src/api/submission.ts
import api from "./axios";

export async function getAssignmentSubmissions(assignmentId: number) {
    const token = localStorage.getItem("token");

    const response = await api.get(
        `/assignments/${assignmentId}/submissions`,
        {
            headers: {
                Authorization: `Bearer ${token}`
            }
        }
    );

    return response.data;
}

export async function gradeSubmission(
    submissionId: number,
    grade: number
) {
    const token = localStorage.getItem("token");

    const response = await api.patch(
        `/submissions/${submissionId}/grade`,
        { grade },
        {
            headers: {
                Authorization: `Bearer ${token}`
            }
        }
    );

    return response.data;
}

export async function submitAssignment(
    assignmentId: number,
    content: string
) {
    const token = localStorage.getItem("token");

    const response = await api.post(
        `/assignments/${assignmentId}/submit`,
        {
            content
        },
        {
            headers: {
                Authorization: `Bearer ${token}`
            }
        }
    );

    return response.data;
}

export async function getMySubmissions() {
    const token = localStorage.getItem("token");

    const response = await api.get(
        "/submissions/me",
        {
            headers: {
                Authorization: `Bearer ${token}`
            }
        }
    );

    return response.data;
}

File: backend/app/api/routes/submission.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.db.database import get_db
from app.models.assignment import Assignment
from app.models.enrollment import Enrollment
from app.models.submission import Submission
from app.models.user import User
from app.schemas.submission import (
    SubmissionCreate,
    SubmissionResponse
)

router = APIRouter(
    tags=["Submissions"]
)


@router.post(
    "/assignments/{assignment_id}/submit",
    response_model=SubmissionResponse,
    status_code=status.HTTP_201_CREATED
)
def submit_assignment(
    assignment_id: int,
    submission_data: SubmissionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    # Only students can submit
    if current_user.role != "student":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only students can submit assignments"
        )

    # Check assignment exists
    assignment = db.query(Assignment).filter(
        Assignment.id == assignment_id
    ).first()

    if not assignment:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Assignment not found"
        )

    # Check student is enrolled in the course
    enrollment = db.query(Enrollment).filter(
        Enrollment.student_id == current_user.id,
        Enrollment.course_id == assignment.course_id
    ).first()

    if not enrollment:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You are not enrolled in this course"
        )

    # Prevent duplicate submissions
    existing_submission = db.query(Submission).filter(
        Submission.student_id == current_user.id,
        Submission.assignment_id == assignment_id
    ).first()

    if existing_submission:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Assignment already submitted"
        )

    # Create submission
    new_submission = Submission(
        assignment_id=assignment_id,
        student_id=current_user.id,
        content=submission_data.content
    )

    db.add(new_submission)
    db.commit()
    db.refresh(new_submission)

    return new_submission


@router.get(
    "/submissions/me",
    response_model=list[SubmissionResponse]
)
def get_my_submissions(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    # Only students can view their submissions
    if current_user.role != "student":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only students can view submissions"
        )

    submissions = db.query(Submission).filter(
        Submission.student_id == current_user.id
    ).all()

    return submissions

@router.get(
    "/assignments/{assignment_id}/submissions",
    response_model=list[SubmissionResponse]
)
def get_assignment_submissions(
    assignment_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    if current_user.role != "teacher":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only teachers can view submissions"
        )
    assignment = db.query(Assignment).filter(
        Assignment.id == assignment_id
    ).first()

    if not assignment:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Assignment not found"
        )
    if assignment.course.teacher_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You cannot view submissions for this assignment"
        )
    submissions = db.query(Submission).filter(
        Submission.assignment_id == assignment_id
    ).all()

    return submissions